src/SFML/Network/TcpSocket.cpp

```diff
@@ -51,16 +74,633 @@ constexpr int flags = MSG_NOSIGNAL;
+                    return TlsStatus::Error;
+                }
+            }
+
+            // Set up how the TLS implementation communicates with the underlying socket
+            state.netContext.fd = static_cast<int>(socket.getNativeHandle());
+
+            mbedtls_ssl_set_bio(
+                &state.sslContext,
+                &socket,
+                [](void* context, const unsigned char* data, std::size_t size) -> int
+                {
+                    auto& tcpSocket = *static_cast<TcpSocket*>(context);
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wuseless-cast"
+                    const auto result = static_cast<int>(
+                        ::send(tcpSocket.getNativeHandle(),
+                               reinterpret_cast<const char*>(data),
+                               static_cast<priv::SocketImpl::Size>(size),
+                               flags));
+
+                    if ((result == -1) && (priv::SocketImpl::getErrorStatus() == sf::Socket::Status::NotReady))
+                        return MBEDTLS_ERR_SSL_WANT_WRITE;
+
+                    return result;
+#pragma GCC diagnostic pop
+                },
+                [](void* context, unsigned char* data, std::size_t size) -> int
+                {
+                    auto& tcpSocket = *static_cast<TcpSocket*>(context);
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wuseless-cast"
+                    const auto result = static_cast<int>(
+                        recv(tcpSocket.getNativeHandle(),
+                             reinterpret_cast<char*>(data),
+                             static_cast<priv::SocketImpl::Size>(size),
+                             flags));
+
+                    if ((result == -1) && (priv::SocketImpl::getErrorStatus() == sf::Socket::Status::NotReady))
+                        return MBEDTLS_ERR_SSL_WANT_READ;
+
+                    return result;
+#pragma GCC diagnostic pop
+                },
+                nullptr);
+        }
+
+        auto& state = *tlsState;
+
+        // Perform the TLS handshake if it isn't complete yet
+        if (!state.handshakeComplete)
+        {
+            if (auto result = mbedtls_ssl_handshake(&state.sslContext); result != 0)
+            {
+                if (result == MBEDTLS_ERR_SSL_WANT_READ || result == MBEDTLS_ERR_SSL_WANT_WRITE)
+                    return TlsStatus::HandshakeStarted;
+
+#if defined(MBEDTLS_ERR_SSL_RECEIVED_NEW_SESSION_TICKET)
+                if (result == MBEDTLS_ERR_SSL_RECEIVED_NEW_SESSION_TICKET)
+                    return TlsStatus::HandshakeStarted;
+#endif
+
+                // Output the reason for verification failure
+                if (result == MBEDTLS_ERR_X509_CERT_VERIFY_FAILED)
+                {
+                    auto verifyResult = mbedtls_ssl_get_verify_result(&state.sslContext);
+
+                    std::string errors;
+
+                    if (verifyResult & MBEDTLS_X509_BADCERT_EXPIRED)
```

Comment:
I'd also remove this bit from `verifyResult`, and then in debug builds you can warn on any bits left. (same for all other cases here of course)